java/src/org/openqa/selenium/grid/node/local/LocalNode.java

```diff
@@ -471,6 +477,49 @@ public HttpResponse executeWebDriverCommand(HttpRequest req) {
     return toReturn;
   }
 
+  @Override
+  public HttpResponse downloadFile(HttpRequest req, SessionId id) {
+    // When the session is running in a Docker container, the download file command
+    // needs to be forwarded to the container as well.
+    SessionSlot slot = currentSessions.getIfPresent(id);
+    if (slot != null && slot.getSession() instanceof DockerSession) {
+      return executeWebDriverCommand(req);
+    }
+    if (this.downloadsDir.isEmpty()) {
+      throw new WebDriverException(
+        "Please specify the directory that would contain downloaded files and restart the node.");
+    }
+    File dir = new File(this.downloadsDir);
+    if (!dir.exists()) {
+      throw new WebDriverException(
+        String.format("Cannot locate downloads directory %s.", downloadsDir));
+    }
+    if (!dir.isDirectory()) {
+      throw new WebDriverException(String.format("Invalid directory: %s.",downloadsDir));
+    }
+    String filename = req.getQueryParameter("filename");
+    try {
+      File[] allFiles = Optional.ofNullable(
+        dir.listFiles((dir1, name) -> name.equals(filename))
+      ).orElse(new File[]{});
+      if (allFiles.length == 0) {
+        throw new WebDriverException(
+          String.format("Cannot find file [%s] in directory %s.", filename, downloadsDir));
+      }
+      if (allFiles.length != 1) {
+        throw new WebDriverException(
+          String.format("Expected there to be only 1 file. There were: %s.", allFiles.length));
+      }
+      String content = Contents.string(allFiles[0]);
+      ImmutableMap<String, Object> result = ImmutableMap.of(
+        "filename", filename,
+        "contents", content);
+      return new HttpResponse().setContent(asJson(result));
```

Comment:
Wouldn't it be better to zip the file and send it back?

Reply:
I did think of that as the first implementation but then again, the user wouldn't know that they are going to get back a zip file which they would need to unzip, even though they are downloading only 1 file at a time. We can perhaps provide an additional query parameter to indicate that the user needs the file in zip format? What do you think ?

Reply:
It is about the traffic the Grid needs to handle, sending several MB files across the Node, Router, etc... vs compressing them to keep the traffic as light as possible. We can always document how they get the file and provide examples on different languages showing how to unzip a file.

Reply:
Ok. Cool. I will change this to accommodate a zip file being sent back.